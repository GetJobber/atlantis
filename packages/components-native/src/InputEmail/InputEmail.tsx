import type { Ref } from "react";
import React, { forwardRef } from "react";
import type { InputTextProps, InputTextRef } from "../InputText";
import { InputText } from "../InputText";

export const InputEmail = forwardRef(InputEmailInternal);
type InputEmailProps = Omit<InputTextProps, "keyboard">;

function InputEmailInternal(props: InputEmailProps, ref: Ref<InputTextRef>) {
  return (
    <InputText
      {...props}
      ref={ref}
      autoCapitalize="none"
      autoCorrect={false}
      keyboard={"email-address"}
      validations={{
        pattern: {
          value:
            /[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?/,
          message: "Enter a valid email address (email@example.com)",
        },
        ...props.validations,
      }}
    />
  );
}
