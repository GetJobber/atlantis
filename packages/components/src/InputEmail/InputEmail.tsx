import React from "react";
import { CommonFormFieldProps, FormField, FormFieldProps } from "../FormField";

type InputEmailProps = CommonFormFieldProps &
  Pick<FormFieldProps, "maxLength" | "readonly" | "validations">;

export const validationMessage = "Please enter a valid email";

export function InputEmail(props: InputEmailProps) {
  const { validations } = props;

  return (
    <FormField
      {...props}
      type="email"
      validations={{
        ...validations,
        validate: checkForValidEmail,
      }}
    />
  );

  function checkForValidEmail(value: string) {
    const emailRegex =
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    if (!value) {
      return true;
    }

    if (!value.match(emailRegex)) {
      return validationMessage;
    }

    return true;
  }
}
