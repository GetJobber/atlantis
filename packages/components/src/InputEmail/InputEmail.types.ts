import { Clearable } from "@jobber/hooks";
import { CommonFormFieldProps, FormFieldProps } from "../FormField";

export type InputEmailLegacyProps = CommonFormFieldProps &
  Pick<
    FormFieldProps,
    "maxLength" | "readonly" | "validations" | "defaultValue"
  >;

export const validationMessage = "Please enter a valid email";

export type InputEmailVersion = 1 | 2 | undefined;

/**
 * Experimental version 2 of the InputEmail component.
 * Do not use unless you have talked with Atlantis first.
 */
export interface InputEmailRebuiltProps
  extends Omit<
    React.InputHTMLAttributes<HTMLInputElement>,
    | "onChange"
    | "onBlur"
    | "maxLength"
    | "size"
    | "suffix"
    | "prefix"
    | "value"
    | "max"
    | "min"
    | "defaultValue"
  > {
  readonly error?: string;
  readonly invalid?: boolean;
  readonly identifier?: string;
  readonly autocomplete?: boolean | string;
  readonly loading?: boolean;
  readonly onKeyUp?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  readonly children?: React.ReactNode;
  readonly clearable?: Clearable;

  /**
   * Version 2 is highly experimental. Avoid using it unless you have talked with Atlantis first.
   */
  readonly version: 2;

  readonly onChange?: (
    newValue: string,
    event?: React.ChangeEvent<HTMLInputElement>,
  ) => void;
  readonly onEnter?: FormFieldProps["onEnter"];
  readonly onBlur?: FormFieldProps["onBlur"];
  readonly value: string;
  readonly maxLength?: number;
  readonly size?: FormFieldProps["size"];
  readonly inline?: FormFieldProps["inline"];
  readonly align?: FormFieldProps["align"];
  readonly toolbar?: FormFieldProps["toolbar"];
  readonly toolbarVisibility?: FormFieldProps["toolbarVisibility"];
  readonly prefix?: FormFieldProps["prefix"];
  readonly suffix?: FormFieldProps["suffix"];
  readonly description?: FormFieldProps["description"];
}

export interface InputEmailRef {
  insert(text: string): void;
  blur(): void;
  focus(): void;
  scrollIntoView(arg?: boolean | ScrollIntoViewOptions): void;
}
