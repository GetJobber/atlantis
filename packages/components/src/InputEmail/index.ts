export { InputEmail, validationMessage } from "./InputEmail";
