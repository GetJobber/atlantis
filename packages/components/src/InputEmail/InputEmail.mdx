---
name: InputEmail
menu: Components
route: /components/input-email
showDirectoryLink: true
---

import { Playground, Props } from "docz";
import { InputEmail } from ".";

# Input Email

InputEmail is used to display an input with built in email validation

```ts
import { InputEmail } from "@jobber/components/InputEmail";
```

<Playground>
  <InputEmail placeholder="Enter your email" />
</Playground>

## Props

<Props of={InputEmail} />

## Usage Guidelines

Use the `<InputEmail>` when you expect the user to input an email address.

If you are not worried about email address validation, consider using
`InputText`.

## Additional validation

By default the `InputEmail` will validate email addresses. You can pass in
additional validations as well.

<Playground>
  <InputEmail
    placeholder="Enter your email"
    validations={{
      required: { value: true, message: "Email address is required" },
    }}
  />
</Playground>
