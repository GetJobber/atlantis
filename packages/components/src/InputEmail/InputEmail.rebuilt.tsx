import React, { forwardRef, useId, useRef } from "react";
import { useInputEmailActions } from "./hooks/useInputEmailActions";
import { useInputEmailFormField } from "./hooks/useInputEmailFormField";
import { InputEmailRebuiltProps } from "./InputEmail.types";
import {
  FormFieldWrapper,
  useAtlantisFormFieldName,
  useFormFieldWrapperStyles,
} from "../FormField";
import { FormFieldPostFix } from "../FormField/FormFieldPostFix";

export const InputEmailRebuilt = forwardRef<
  HTMLInputElement,
  InputEmailRebuiltProps
>(
  (
    {
      error,
      invalid,
      identifier,
      autocomplete,
      loading,
      onKeyDown,
      children,
      onChange,
      onEnter,
      onBlur,
      onFocus,
      value,
      maxLength,
      size,
      inline,
      align,
      toolbar,
      toolbarVisibility,
      prefix,
      suffix,
      description,
      ...props
    },
    ref,
  ) => {
    const id = useId();
    const inputRef = useRef<HTMLInputElement>(null);
    const wrapperRef = useRef<HTMLDivElement>(null);

    const { inputStyle } = useFormFieldWrapperStyles({
      size,
      inline,
      align,
      type: "text",
      value,
      invalid,
      error,
      maxLength,
      disabled: props.disabled,
    });

    const { name } = useAtlantisFormFieldName({
      nameProp: props.name,
      id,
    });

    const {
      handleChange,
      handleBlur,
      handleFocus,
      handleKeyDown,
      handleClear,
    } = useInputEmailActions({
      onChange,
      onBlur,
      onFocus,
      onKeyDown,
      onEnter,
      inputRef,
    });

    const { fieldProps, descriptionIdentifier } = useInputEmailFormField({
      ...props,
      id,
      name,
      handleChange,
      handleBlur,
      handleFocus,
      handleKeyDown,
    });

    return (
      <FormFieldWrapper
        error={error || ""}
        invalid={invalid}
        identifier={identifier || id}
        descriptionIdentifier={descriptionIdentifier}
        size={size}
        inline={inline}
        align={align}
        toolbar={toolbar}
        toolbarVisibility={toolbarVisibility}
        prefix={prefix}
        suffix={suffix}
        description={description}
        clearable={props.clearable ?? "never"}
        onClear={handleClear}
        wrapperRef={wrapperRef}
        maxLength={maxLength}
        disabled={props.disabled}
        type="text"
        value={value}
      >
        <input
          {...fieldProps}
          ref={ref || inputRef}
          type="email"
          className={inputStyle}
          value={value}
          autoComplete={autocomplete === true ? "email" : autocomplete || "off"}
          aria-invalid={invalid}
          aria-describedby={error ? `${identifier || id}-error` : undefined}
        />
        <FormFieldPostFix variation="spinner" visible={loading ?? false} />
        {children}
      </FormFieldWrapper>
    );
  },
);

InputEmailRebuilt.displayName = "InputEmailRebuilt";
