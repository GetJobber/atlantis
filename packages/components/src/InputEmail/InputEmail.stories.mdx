import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { InputEmail } from ".";

<Meta title="Components/Forms and Inputs/InputEmail" component={InputEmail} />

# Input Email

InputEmail is used to display input with built-in email validation.

```ts
import { InputEmail } from "@jobber/components/InputEmail";
```

<Canvas>
  <Story name="InputEmail" args={{ placeholder: "Enter your email" }}>
    {args => <InputEmail {...args} />}
  </Story>
</Canvas>

## Props

<ArgsTable of={InputEmail} story="InputEmail" />

## Usage guidelines

Use the `<InputEmail>` when you expect the user to input an email address.

If you are not worried about email address validation, consider using
`InputText`.

## Additional validation

By default, the `InputEmail` will validate email addresses. You can pass in
additional validations as well.

<Canvas>
  <InputEmail
    placeholder="Enter your email"
    validations={{
      required: { value: true, message: "Email address is required" },
    }}
  />
</Canvas>
