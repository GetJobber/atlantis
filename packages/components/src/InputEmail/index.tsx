import React, { ForwardedRef, forwardRef } from "react";
import { InputEmail as InputEmailLegacy } from "./InputEmail";
import { InputEmailRebuilt } from "./InputEmail.rebuilt";
import {
  type InputEmailLegacyProps,
  type InputEmailRebuiltProps,
} from "./InputEmail.types";

export type InputEmailShimProps =
  | InputEmailRebuiltProps
  | InputEmailLegacyProps;

function isNewInputEmailProps(
  props: InputEmailShimProps,
): props is InputEmailRebuiltProps {
  return props.version === 2;
}

export const InputEmail = forwardRef(function InputEmailShim(
  props: InputEmailShimProps,
  ref: ForwardedRef<HTMLInputElement>,
) {
  if (isNewInputEmailProps(props)) {
    return (
      <InputEmailRebuilt
        {...props}
        ref={ref as ForwardedRef<HTMLInputElement>}
      />
    );
  } else {
    return <InputEmailLegacy {...props} />;
  }
});
export { InputEmailRebuiltProps, InputEmailLegacyProps };
