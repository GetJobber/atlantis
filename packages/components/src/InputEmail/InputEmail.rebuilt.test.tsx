import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { InputEmailRebuilt } from "./InputEmail.rebuilt";

describe("InputEmailRebuilt", () => {
  it("renders a basic InputEmailRebuilt", () => {
    const { container } = render(<InputEmailRebuilt version={2} />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly with placeholder", () => {
    const { container } = render(
      <InputEmailRebuilt version={2} placeholder="Enter email" />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly as small", () => {
    const { container } = render(
      <InputEmailRebuilt version={2} size="small" />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly as large", () => {
    const { container } = render(
      <InputEmailRebuilt version={2} size="large" />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when disabled", () => {
    const { container } = render(<InputEmailRebuilt version={2} disabled />);
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when invalid", () => {
    const { container } = render(<InputEmailRebuilt version={2} invalid />);
    expect(container).toMatchSnapshot();
  });

  it("renders the value when set", () => {
    const value = "test@example.com";
    render(<InputEmailRebuilt version={2} value={value} />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveValue(value);
  });

  describe("Action handlers", () => {
    it("should call onChange with the new value when input changes", async () => {
      const changeHandler = jest.fn();
      const initialValue = "initial@example.com";
      const newValue = "new@example.com";

      render(
        <InputEmailRebuilt
          version={2}
          value={initialValue}
          onChange={changeHandler}
        />,
      );

      const inputElement = screen.getByRole("textbox");
      fireEvent.change(inputElement, { target: { value: newValue } });
      expect(changeHandler).toHaveBeenCalledWith(newValue, expect.any(Object));
    });

    it("should call onFocus when the input is focused", async () => {
      const focusHandler = jest.fn();
      const value = "test@example.com";

      render(
        <InputEmailRebuilt version={2} value={value} onFocus={focusHandler} />,
      );

      await userEvent.click(screen.getByDisplayValue(value));
      expect(focusHandler).toHaveBeenCalledTimes(1);
    });

    it("should call onBlur when the input loses focus", async () => {
      const blurHandler = jest.fn();

      render(
        <>
          <InputEmailRebuilt version={2} onBlur={blurHandler} />
          <button type="button" data-testid="other-element">
            Other element
          </button>
        </>,
      );

      const inputElement = screen.getByRole("textbox");
      await userEvent.click(inputElement);
      await userEvent.click(screen.getByTestId("other-element"));

      expect(blurHandler).toHaveBeenCalledTimes(1);
    });

    it("should call onChange with undefined, call onBlur, and focus input when cleared", async () => {
      const changeHandler = jest.fn();
      const blurHandler = jest.fn();
      const ref = React.createRef<HTMLInputElement>();
      const initialValue = "test@example.com";

      render(
        <InputEmailRebuilt
          version={2}
          value={initialValue}
          onChange={changeHandler}
          onBlur={blurHandler}
          clearable="always"
          ref={ref}
        />,
      );

      const clearButton = screen.getByTestId("ATL-FormField-clearButton");
      await userEvent.click(clearButton);

      expect(changeHandler).toHaveBeenCalledWith("");
      expect(blurHandler).toHaveBeenCalledTimes(1);
    });
  });

  describe("Ref handling", () => {
    it("should forward ref to the input element", () => {
      const ref = React.createRef<HTMLInputElement>();

      render(<InputEmailRebuilt version={2} ref={ref} />);

      expect(ref.current).toBeInstanceOf(HTMLInputElement);
    });

    it("should allow access to native input methods through ref", async () => {
      const ref = React.createRef<HTMLInputElement>();

      render(<InputEmailRebuilt version={2} ref={ref} />);

      const inputElement = screen.getByRole("textbox");
      const focusSpy = jest.spyOn(inputElement, "focus");
      const blurSpy = jest.spyOn(inputElement, "blur");

      ref.current?.focus();
      expect(focusSpy).toHaveBeenCalled();

      ref.current?.blur();
      expect(blurSpy).toHaveBeenCalled();
    });
  });
});
