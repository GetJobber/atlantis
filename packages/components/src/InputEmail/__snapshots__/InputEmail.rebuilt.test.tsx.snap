// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputEmailRebuilt renders a basic InputEmailRebuilt 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              id=":r0:"
              name="generatedName--:r0:"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputEmailRebuilt renders correctly as large 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper large"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              id=":r3:"
              name="generatedName--:r3:"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputEmailRebuilt renders correctly as small 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper small"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              id=":r2:"
              name="generatedName--:r2:"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputEmailRebuilt renders correctly when disabled 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper disabled"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              disabled=""
              id=":r4:"
              name="generatedName--:r4:"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputEmailRebuilt renders correctly when invalid 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper invalid"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              id=":r5:"
              invalid="true"
              name="generatedName--:r5:"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`InputEmailRebuilt renders correctly with placeholder 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="wrapper"
      data-testid="Form-Field-Wrapper"
    >
      <div
        class="horizontalWrapper"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="childrenWrapper"
            tabindex="-1"
          >
            <input
              class="input"
              data-testid="ATL-InputEmail-input"
              id=":r1:"
              name="generatedName--:r1:"
              placeholder="Enter email"
              type="email"
              value=""
              version="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
